test: Add swim proto fuzz targets.

SweetVishnya opened the pull request:
This PR moves swim fuzz targets to Tarantool repo.

ref #6614

SweetVishnya commented:
Something is wrong with test-run submodules. Do you have any ideas?

Totktonada commented:
> Unable to find current revision in submodule path 'test-run/lib/tarantool-python'

But current submoduel commit is reachable from master: https://github.com/tarantool/tarantool-python/commit/ea942f981158a3dcf7d5d343ad3acf89edd4b15b. Strange.

@ylobankov Can it be a local problem with self-hosted runners (something about submodules update)?

Totktonada commented:
I don't know how the repository on the machine in CI pool got this state, but `test-run/lib/tarantool-python` had only `.git` file with `gitdir: <...>`. `git pull` in the directory fixes it. Restarted jobs.

MacOS 11.0 fail is due to #6576.

ligurio reviewed (changes requested):
Thanks for your fuzzers!

I build both fuzzers on my local machine with ASAN (our OSS Fuzz config uses UBSAN and ASAN [1]) and see that both fuzzers failed with SEGV
```
==1153379==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x000000566fa6 bp 0x7ffcc0883110 sp 0x7ffcc08830f0 T0)
==1153379==The signal is caused by a READ memory access.
==1153379==Hint: address points to the zero page.
    #0 0x566fa6 in diag_get /home/sergeyb/sources/MRG/tarantool/src/lib/core/diag.c:146:21
    #1 0x55d530 in swim_decode_map /home/sergeyb/sources/MRG/tarantool/src/lib/swim/swim_proto.c:56:3
    #2 0x55f18b in swim_member_def_decode /home/sergeyb/sources/MRG/tarantool/src/lib/swim/swim_proto.c:308:6
```

here [2]:

```c
143 struct diag *                                                                    
144 diag_get(void)                                                                   
145 {                                                                                
146         rerturn &fiber()->diag;                                                  
147 }                                                                                
148   
```
because `fiber()` and `cord()` are macro and `cord_ptr` is undefined in our case [3]:
```c
extern __thread struct cord *cord_ptr;

#define cord() cord_ptr
#define fiber() cord()->fiber
#define loop() (cord()->loop)
``` 

We should fix this before merging to master because fuzzing is a part of our CI and CI will be red on next run of fuzzing in ClusterFuzz.

1. https://github.com/google/oss-fuzz/blob/master/projects/tarantool/project.yaml#L11-L13
2. https://github.com/tarantool/tarantool/blob/master/src/lib/core/diag.c#L143-L147
3. https://github.com/tarantool/tarantool/blob/master/src/lib/core/fiber.h#L733-L737

ligurio commented:
@SweetVishnya Also we have a corpus for each fuzzer for more efficient fuzzing, see [test/static/corpus](https://github.com/tarantool/tarantool/tree/master/test/static/corpus). Will you put corpus for SWIM fuzzers too?

SweetVishnya commented:
Yeah, I can provide corpus from fuzzer.

Fuzzers actually crash because they need [this](https://github.com/ispras/oss-sydr-fuzz/blob/master/projects/tarantool/disable_logging.diff) patch to disable logging. Do you have a propper way to do it?

SweetVishnya commented:
Yeah, I can provide corpus from fuzzer.

Fuzzers actually crash because they need [this](https://github.com/ispras/oss-sydr-fuzz/blob/master/projects/tarantool/disable_logging.diff) patch to disable logging. Do you have a propper way to do it?

ligurio commented:
> Yeah, I can provide corpus from fuzzer.

It would be nice!

> Fuzzers actually crash because they need [this](https://github.com/ispras/oss-sydr-fuzz/blob/master/projects/tarantool/disable_logging.diff) patch to disable logging. Do you have a propper way to do it?

I believe that @Gerold103 can help with this.

Gerold103 commented:
After > 30 minutes I failed to install libfuzzer - it keeps complaining about missing fuzzer library during linking. Installation from scratch took > 15 minutes and then failed with some error about XCode. So I give up to run it locally. Can try if anybody would show me a docker command to run it without troubles.

But I can say even now that you need to initialize fiber module. Look at `swim_run_test()` how it is done. You would need to do this in the beginning:
```C
	random_init();
	memory_init();
	fiber_init(fiber_c_invoke);
	say_logger_init(...);
```
And this in the end:
```C
	say_logger_free();
	fiber_free();
	memory_free();
	random_free();
```
But you probably need not all of that. Try to experiment with removing some of these lines and see if it works.

SweetVishnya commented:
@Gerold103, libFuzzer is easy to install. It is a part of llvm (clang). So, just installing clang works fine.

Thank you for the initiallization code. That is just what I was looking for. I'll try to fix logging with that code.

Gerold103 commented:
I have clang. And I even upgraded it. But the libfuzzer wasn't installed with it somewhy. I am getting this error: `ld: file not found: /Library/Developer/CommandLineTools/usr/lib/clang/12.0.0/lib/darwin/libclang_rt.fuzzer_osx.a` no matter what I do. And that file indeed does not exist. As well as any file having 'fuzzer' in its name in `/Library/Developer/CommandLineTools/usr/lib/`.

SweetVishnya commented:
Hmm, seems to be some specific osx issue. clang 12 should provide libFuzzer. Maybe it is in a separate package for Mac OS. Anyway, building in ubuntu 20 [docker](https://github.com/ispras/oss-sydr-fuzz/tree/master/projects/tarantool) works fine.

P.S. I am working on fiber init right now.

SweetVishnya commented:
@ligurio, I modified fuzz targets. Now everything works fine with current OSS-Fuzz without patch. Also, I supplied corpus seeds for the targets. This PR is ready for merge.

ligurio commented:
@SweetVishnya thanks for changes and added corpus!

swim_proto_member_fuzzer failed:
```
$ cmake -DENABLE_FUZZER=ON -DENABLE_ASAN=ON -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER="/usr/bin/clang" -DCMAKE_CXX_COMPILER="/usr/bin/clang++" ..
$ make -j
$ ./test/fuzz/swim_proto_member_fuzzer -workers=4 ../test/static/corpus/swim_proto_member/
...
#75155  REDUCE cov: 299 ft: 429 corp: 72/1338b lim: 298 exec/s: 6262 rss: 410Mb L: 12/240 MS: 3 ChangeBinInt-ChangeByte-EraseBytes-
#75736  REDUCE cov: 299 ft: 429 corp: 72/1337b lim: 298 exec/s: 6311 rss: 411Mb L: 11/240 MS: 1 EraseBytes-       
swim_proto_member_fuzzer: /home/sergeyb/sources/MRG/tarantool/src/lib/msgpuck/msgpuck.h:2052: ptrdiff_t mp_check_binl(const char *, const char *): Assertion `
cur < end' failed.                                                             
==1215751== ERROR: libFuzzer: deadly signal                                                                                                                   
    #0 0x5338e1 in __sanitizer_print_stack_trace (/home/sergeyb/sources/MRG/tarantool/build/test/fuzz/swim_proto_member_fuzzer+0x5338e1)
    #1 0x47ea38 in fuzzer::PrintStackTrace() (/home/sergeyb/sources/MRG/tarantool/build/test/fuzz/swim_proto_member_fuzzer+0x47ea38)
    #2 0x463b83 in fuzzer::Fuzzer::CrashCallback() (/home/sergeyb/sources/MRG/tarantool/build/test/fuzz/swim_proto_member_fuzzer+0x463b83)
    #3 0x7f094dfc43bf  (/lib/x86_64-linux-gnu/libpthread.so.0+0x153bf)
    #4 0x7f094ba8518a in __libc_signal_restore_set /build/glibc-eX1tMB/glibc-2.31/signal/../sysdeps/unix/sysv/linux/internal-signals.h:86:3
    #5 0x7f094ba8518a in raise /build/glibc-eX1tMB/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:48:3
    #6 0x7f094ba64858 in abort /build/glibc-eX1tMB/glibc-2.31/stdlib/abort.c:79:7
    #7 0x7f094ba64728 in __assert_fail_base /build/glibc-eX1tMB/glibc-2.31/assert/assert.c:92:3
    #8 0x7f094ba75f35 in __assert_fail /build/glibc-eX1tMB/glibc-2.31/assert/assert.c:101:3
    #9 0x63ba3f in mp_check_binl /home/sergeyb/sources/MRG/tarantool/src/lib/msgpuck/msgpuck.h:2052:2
    #10 0x55ea79 in swim_decode_bin /home/sergeyb/sources/MRG/tarantool/src/lib/swim/swim_proto.c:129:6
    #11 0x56022e in swim_decode_member_key /home/sergeyb/sources/MRG/tarantool/src/lib/swim/swim_proto.c:287:7
    #12 0x55f631 in swim_member_def_decode /home/sergeyb/sources/MRG/tarantool/src/lib/swim/swim_proto.c:320:7
    #13 0x55d049 in LLVMFuzzerTestOneInput /home/sergeyb/sources/MRG/tarantool/test/fuzz/swim_proto_member_fuzzer.c:41:5
    #14 0x465241 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/sergeyb/sources/MRG/tarantool/build/test/fuzz/swim_proto_member_fuzzer+0x465241)
    #15 0x464985 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) (/home/sergeyb/sources/MRG/tarantool/build/test/fuzz/swim_proto_member_fuzzer+0x464985)
    #16 0x466c27 in fuzzer::Fuzzer::MutateAndTestOne() (/home/sergeyb/sources/MRG/tarantool/build/test/fuzz/swim_proto_member_fuzzer+0x466c27)
    #17 0x467925 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) (/home/sergeyb/sources/MRG/tarantool/build/test/fuzz/swim_proto_member_fuzzer+0x467925)
    #18 0x4562de in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/sergeyb/sources/MRG/tarantool/build/test/fuzz/swim_proto_member_fuzzer+0x4562de)
    #19 0x47f122 in main (/home/sergeyb/sources/MRG/tarantool/build/test/fuzz/swim_proto_member_fuzzer+0x47f122)
    #20 0x7f094ba660b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/../csu/libc-start.c:308:16
    #21 0x42b07d in _start (/home/sergeyb/sources/MRG/tarantool/build/test/fuzz/swim_proto_member_fuzzer+0x42b07d)

NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 1 EraseBytes-; base unit: 12af317b9a79dfbb252705a3bf1bd6b35b18f192
0x8a,0x6,0xc5,
\x8a\x06\xc5
artifact_prefix='./'; Test unit written to ./crash-3e7586574c3ce7572c7e3740edd074d357478b4f
Base64: igbF
```

ligurio commented:
@Gerold103, you can follow the [guide for OSS-Fuzz](https://google.github.io/oss-fuzz/getting-started/new-project-guide/#testing-locally) or steps below to run fuzzers with Tarantool:

```sh
$ git clone https://github.com/google/oss-fuzz
$ cd oss-fuzz/
$ git diff
--- a/projects/tarantool/Dockerfile
+++ b/projects/tarantool/Dockerfile
@@ -20,7 +20,8 @@ RUN apt-get update && apt-get install -y \
        autoconf automake libtool zlib1g-dev \
        libreadline-dev libncurses5-dev libssl-dev \
        libunwind-dev libicu-dev luajit
-RUN git clone https://github.com/tarantool/tarantool
+RUN git clone https://github.com/SweetVishnya/tarantool
 WORKDIR tarantool
+RUN git checkout fuzz-swim
 RUN git submodule update --init --recursive
 COPY build.sh $SRC/
$ python infra/helper.py pull_images
$ python infra/helper.py build_image tarantool
$ python infra/helper.py build_fuzzers --sanitizer [address | undefined] --architecture x86_64 tarantool
$ python infra/helper.py run_fuzzer --engine libfuzzer --sanitizer address tarantool swim_proto_meta_fuzzer
$ 
```

SweetVishnya commented:
I reproduced the crash. Looking into it. Seems to be some debug assertions that fail. May be a possible bug or not.

SweetVishnya commented:
Yeah, this is a bug indeed.

`cur == end` after `mp_load_u8`:

```
│   2048        MP_IMPL ptrdiff_t                                                                                                                                                             │
│   2049        mp_check_binl(const char *cur, const char *end)                                                                                                                               │
│   2050        {                                                                                                                                                                             │
│   2051                uint8_t c = mp_load_u8(&cur);                                                                                                                                         │
│  >2052                assert(cur < end);                                                                                                                                                    │
│   2053                assert(mp_typeof(c) == MP_BIN);                                                                                                                                       │
│   2054                                                                                                                                                                                      │
│   2055                assert(c >= 0xc4 && c <= 0xc6); /* must be checked above by mp_typeof */                                                                                              │
│   2056                uint32_t hsize = 1U << (c & 0x3); /* 0xc4->1, 0xc5->2, 0xc6->4 */                                                                                                     │
│   2057                return hsize - (end - cur);                                                                                                                                           │
│   2058        } 
```

SweetVishnya commented:
The main problem is that errors are handled via assert, not the return value.

SweetVishnya commented:
Working on a fix.

Gerold103 commented on test/fuzz/swim_proto_member_fuzzer.c:
```diff
@@ -0,0 +1,40 @@
+
+void cord_on_yield() {}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
```
Please,
* write return type on a separate line;
* write `{` and `}` on their own separate lines.
To conform with the code style.

The same in the other file.

Gerold103 commented on test/fuzz/swim_proto_member_fuzzer.c:
```diff
@@ -0,0 +1,40 @@
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
+{
+    struct swim_member_def def;
+    char *buf = malloc(size);
```
Why do you need to copy the original buffer? Decoding does not change it.

Gerold103 reviewed (changes requested):
Thanks for the fixes! Please, merge the second commit into the first one.

Also the question is - how did you generate the files in the third commit? Are they just garbage or do they contain something meaningful?

SweetVishnya commented:
Fixed review issues, squashed commits, rebased on master (now fuzz targets do not crash).

@Gerold103, I gathered minimized (in terms of coverage) corpus from libFuzzer run.

ligurio commented on test/fuzz/CMakeLists.txt:
```diff
@@ -38,9 +38,17 @@ target_link_libraries(uri_fuzzer PUBLIC uri fuzzer_config)
 set(fuzzing_binaries csv_fuzzer
                      http_parser_fuzzer
-                     uri_fuzzer)
+                     uri_fuzzer
+                     swim_proto_member_fuzzer
+                     swim_proto_meta_fuzzer)
```
We can keep an alphabetical order here.

SweetVishnya replied:
Done

ligurio reviewed (changes requested):
Thanks for the patch! LGTM
